neps/nep-0399.md

```diff
@@ -0,0 +1,372 @@
+to calculate the TTN cost for the write. Such reads cannot be replaced by a read in FlatStorage because FlatStorage does
+not provide the path to the trie node. 
+
+There are multiple proposals on how storage writes can work with FlatStorage. 
+- Keep it the same. The cost of writes remain the same. Note that this can increase the cost for writes in 
+  some cases, for example, if a contract first read from a key and then writes to the same key in the same chunk.
+  Without FlatStorage, the key will be cached in the chunk cache after the read, so the write will cost less. 
+  With FlatStorage, the read will go through FlatStorage, the write will not find the key in the chunk cache and 
+  it will cost more.
+- Remove the TTN cost from storage write fees. Currently, there are two ideas in this direction.
+    - Charge based on maximum depth of a contract’s state, instead of per-touch-trie node.
+    - Charge based on key length only.
+      
+      Both of the above ideas would allow us to remove writes from the critical path of block execution. However, 
+    it is unclear at this point what the new cost would look like and whether further optimizations are needed
+      to bring down the cost for writes in the new cost model.
+
+### Migration Plan
+There are two main questions regarding to how to enable FlatStorage.
+1) Whether there should be database migration. The main challenge of enabling FlatStorage will be to build the flat state
+   column, which requires iterating the entire state. We currently estimate that it takes 1 hour to build
+   flat state for archival nodes and 15 minutes for rpc and validator nodes. Note that this estimation is very rough
+   and further verification is needed. The main concern is that if it takes too long for archival node to migrate,
+   they may have a hard time catching up later since the block processing speed of archival nodes is not very fast. 
+
+   Alternatively, we can build the flat state in a background process while the node is running. This provides a better 
+   experience for both archival and validator nodes since the migration process is transient to them. It would require 
+   more implementation effort from our side. 
+   
+   To make a decision, we will verify the time it takes to build flat state. If it will cause a problem for archival nodes
+   to catch up, we will implement the background migration process.
+2) Whether there should be a protocol upgrade. The enabling of FlatStorage itself does not require a protocol upgrade, since
+it is an internal storage implementation that doesn't change protocol level. However, a protocol upgrade is needed 
+   if we want to adjust fees based on the storage performance with FlatStorage. These two changes can happen in one release,
+   or we can be release them separately. We propose that the enabling of FlatStorage and the protocol upgrade 
```

Comment:
FS (for reads) and fee changes have to be updated at the same time, because FS in fact eliminates TTN and we need to compensate for that. The idea is to use combination of storage read base and byte costs AFAIK.